### PR TITLE
Hashmap implementation

### LINKED ISSUE #15: Hashmap
We need some hash map utilities. Probably have the keys be `const char *`s and the values be `void *`. Here is a suggested API:
Creating and destroying:
`int hm_init(Hashmap *hm, int num_buckets);`
`int hm_free(Hashmap *hm);`
Looking up and setting values:
`void* hm_lookup(Hashmap *hm, const char* key);`
`int hm_set(Hashmap *hm, const char *key, void *value);`
Alter this as you see fit, as long as we can use `const char*` keys to access `void*` values and it mostly runs in O(1).

### PR DESCRIPTION
**Pull request by NicoBliss**
should close issue #15

### CODE REVIEW

**Review by adamhutchings — CHANGES_REQUESTED** [review]
Two small-ish things:
For consistency, could we name the structs `HashNode` and `HashMap`
And could you add a test file for this?

But aside from that, looks good!

**Comment by NicoBliss** [comment]
> Two small-ish things: For consistency, could we name the structs `HashNode` and `HashMap` And could you add a test file for this?
> 
> But aside from that, looks good!

All of this has been added (and a bunch of files header's lightly modified... sorry about that)

**Comment by JakeRoggenbuck** [comment]
I get an error when building.
```
.../jccc/src/util/test_hash.c: In function ‘test_hashing’:
.../jccc/src/util/test_hash.c:25:24: error: initialization of ‘const char *’ from incompatible pointer type ‘char (*)[6]’ [-Wincompatible-pointer-types]
   25 |     const char *key1 = &string1;
      |                        ^
.../jccc/src/util/test_hash.c:26:24: error: initialization of ‘const char *’ from incompatible pointer type ‘char (*)[6]’ [-Wincompatible-pointer-types]
   26 |     const char *key2 = &string2;
      |                        ^
```

File: src/util/test_hash.c [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+int test_minimal_case() {
+    testing_func_setup();
+
+    HashMap* table = hm_init(1);
+    const char *key = "hello";
+    void *value = (void *)128;
+    
+    hm_set(table, key, value);
+    tassert(hm_lookup(table, key) == (void *)128);
```
Comment (JakeRoggenbuck) [inline src/util/test_hash.c:51]: You may also want to change this to something like
```c
    tassert(*(int*)hm_lookup(table, key) == 128);
```

**Comment by JakeRoggenbuck** [comment]
When I added `test_minimal_case();` to test_hash

```c
--- a/src/util/test_hash.c
+++ b/src/util/test_hash.c
@@ -10,6 +10,7 @@ int test_hash() {

     test_hashing();
     test_init_and_free();
+       test_minimal_case();
     test_100_inserts();
     test_100_removes();
```

When it's run, it segfaults,
```c
Running "test_hashing"
Running "test_init_and_free"
Running "test_minimal_case"
Warning: jccc: hashmap full!
Segmentation fault (core dumped)
```

**Comment by adamhutchings** [comment]
Closing due to inactivity.

**Pull request closed without merging**